File: solutions/javascript/meetup/1/meetup.js
const START_DATES = {teenth: 13, first: 1, second: 8, third: 15, fourth: 22, last: -6}
const WEEKDAYS = {Sunday: 0, Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6}

export const meetup = (year, month, week, weekday) => {
  const meetupDate = new Date(year, week !== 'last' ? month - 1 : month, START_DATES[week])
  meetupDate.setDate(meetupDate.getDate() + (WEEKDAYS[weekday] - meetupDate.getDay() + 7) % 7)
  return meetupDate
}

File: solutions/haskell/meetup/1/src/Meetup.hs
module Meetup (Weekday(..), Schedule(..), meetupDay) where

import Data.Time.Calendar

data Weekday = Monday
             | Tuesday
             | Wednesday
             | Thursday
             | Friday
             | Saturday
             | Sunday
             deriving Enum

data Schedule = First
              | Second
              | Third
              | Fourth
              | Last
              | Teenth

meetupDay :: Schedule -> Weekday -> Integer -> Int -> Day
meetupDay s wd y m = firstDayOfWeekOnAfter (toEnum $ fromEnum wd + 1) (YearMonthDay y m d)
  where
    d = case s of
      First -> 1
      Second -> 8
      Third -> 15
      Fourth -> 22
      Last -> gregorianMonthLength y m - 6
      Teenth -> 13